MCO-1009: MCO-1008: MCO-905: Implement NodeDisruptionPolicy API

djoshy opened the pull request:
### This PR does the following:
- The operator now checks the MachineConfiguration global knob for any user provided node disruption policies, merges them with cluster defaults and displays the results in the Status of the MachineConfiguration object. 
- The daemon looks at the status of the MachineConfiguration object during node updates. If the daemon is unable to find the status of MachineConfiguration for more than 2 minutes, it will default to the legacy update path.
- All of this is gated on NodeDisruption feature gate

### How to test:

- Bring up a cluster in TechPreview. If testing on a cluster that was originally having the default FeatureSet and then transitioned to TechPreview, please read this https://github.com/openshift/machine-config-operator/pull/4267#issuecomment-2071154702
- Check the status of the MachineConfiguration object named "cluster"; this should list the cluster default policies.
```
$ oc get MachineConfiguration/cluster -o yaml
apiVersion: operator.openshift.io/v1
kind: MachineConfiguration
metadata:
  creationTimestamp: "2024-04-16T15:02:37Z"
  generation: 4
  name: cluster
  resourceVersion: "261205"
  uid: 2c67b155-1898-452f-adbd-ed376afc0ea2
spec:
  logLevel: Normal
  managementState: Managed
  operatorLogLevel: Normal
status:
  nodeDisruptionPolicyStatus:
    clusterPolicies:
      files:
      - actions:
        - type: None
        path: /etc/mco/internal-registry-pull-secret.json
      - actions:
        - type: None
        path: /var/lib/kubelet/config.json
      - actions:
        - reload:
            serviceName: crio.service
          type: Reload
        path: /etc/machine-config-daemon/no-reboot/containers-gpg.pub
      - actions:
        - reload:
            serviceName: crio.service
          type: Reload
        path: /etc/containers/policy.json
      - actions:
        - type: Special
        path: /etc/containers/registries.conf
      sshkey:
        actions:
        - type: None
  readyReplicas: 0

```
- Apply a MachineConfiguration named "cluster" with a valid NodeDisruptionPolicy. More information about the policy and possible actions can be found [here](https://github.com/openshift/enhancements/blob/0264993edbb3ffdb66eada258a77a81f033b7ac1/enhancements/machine-config/admin-defined-node-disruption-policy.md#proposal). Here is a sample policy:
```
apiVersion: operator.openshift.io/v1
kind: MachineConfiguration
metadata:
  name: cluster
  namespace: openshift-machine-config-operator
spec:
  nodeDisruptionPolicy:
    files:
      - path: "/etc/my-file"
        actions:
          - type: None
    units:
      - name: "my.service"
        actions:
          - type: Restart
            restart:
              serviceName: crio.service
      - name: "test.service"
        actions:
          - type: DaemonReload
          - type: Drain
          - type: Reload
            reload:
              serviceName: crio.service
          - type: Restart
            restart:
              serviceName: my.service
    sshkey:
      actions:
      - type: Reboot
```

- Check the status of the MachineConfiguration object; this should now list a merged policy, including the ones specifies by you and the defaults. If there are any conflicts, the user specified ones will override the cluster defaults. 
```

status:
  nodeDisruptionPolicyStatus:
    clusterPolicies:
      files:
      - actions:
        - type: None
        path: /etc/mco/internal-registry-pull-secret.json
      - actions:
        - type: None
        path: /var/lib/kubelet/config.json
      - actions:
        - reload:
            serviceName: crio.service
          type: Reload
        path: /etc/machine-config-daemon/no-reboot/containers-gpg.pub
      - actions:
        - reload:
            serviceName: crio.service
          type: Reload
        path: /etc/containers/policy.json
      - actions:
        - type: Special
        path: /etc/containers/registries.conf
      - actions:
        - type: None
        path: /etc/my-file
      sshkey:
        actions:
        - type: Reboot
      units:
      - actions:
        - restart:
            serviceName: crio.service
          type: Restart
        name: my.service
      - actions:
        - type: DaemonReload
        - type: Drain
        - reload:
            serviceName: crio.service
          type: Reload
        - restart:
            serviceName: my.service
          type: Restart
        name: test.service
```
In this case, the SSHkey section has been overriden. 
- Now, apply a new MachineConfig with a change that would be in effect for the currently defined node disruption policy. In this case, I applied a config which had changes to the `test.service` unit. Observe the daemon logs on the targeted node, and you should see the daemon step through the designated actions. 
```
I0417 18:20:47.614397    2551 update.go:2578] Starting update from rendered-infra-5d1e3cebfddcee59ac7bde56152e0919 to rendered-infra-a17773854499b91cb5cc86087b54cfe8: &{osUpdate:false kargs:false fips:false passwd:false files:false units:true kernelType:false extensions:false}
I0417 18:20:47.624217    2551 update.go:717] Calculating node disruption actions
I0417 18:20:47.624251    2551 update.go:642] NodeDisruptionPolicy found for diff unit test.service!
I0417 18:20:47.624261    2551 update.go:1023] Calculated node disruption actions:
I0417 18:20:47.624271    2551 update.go:1030] DaemonReload
I0417 18:20:47.624284    2551 update.go:1030] Drain
I0417 18:20:47.624298    2551 update.go:1026] Reload - crio.service
I0417 18:20:47.624313    2551 update.go:1028] Restart - my.service
I0417 18:20:47.624320    2551 drain.go:121] Checking drain required for node disruption actions
I0417 18:20:47.624328    2551 update.go:1060] Drain calculated for node disruption: true
...
...
...
I0417 18:21:30.264732    2551 update.go:2578] Executing performPostConfigChangeNodeDisruptionAction(drain already complete/skipped) for config rendered-infra-a17773854499b91cb5cc86087b54cfe8
I0417 18:21:30.266752    2551 update.go:2578] Executing postconfig action: DaemonReload for config rendered-infra-a17773854499b91cb5cc86087b54cfe8
I0417 18:21:30.268793    2551 update.go:2563] Running: systemctl daemon-reload
I0417 18:21:30.645513    2551 update.go:2578] daemon-reload service reloaded successfully!
I0417 18:21:30.647826    2551 update.go:2578] Executing postconfig action: Drain for config rendered-infra-a17773854499b91cb5cc86087b54cfe8
I0417 18:21:30.649623    2551 update.go:2578] Executing postconfig action: Reload for config rendered-infra-a17773854499b91cb5cc86087b54cfe8
I0417 18:21:30.651303    2551 update.go:2563] Running: systemctl reload crio.service
I0417 18:21:30.688445    2551 update.go:2578] crio.service service reloaded successfully!
I0417 18:21:30.690534    2551 update.go:2578] Executing postconfig action: Restart for config rendered-infra-a17773854499b91cb5cc86087b54cfe8
I0417 18:21:30.692379    2551 update.go:2563] Running: systemctl restart my.service
I0417 18:21:30.711817    2551 update.go:2578] my.service service restarted successfully!
I0417 18:21:30.720163    2551 daemon.go:1574] Previous boot ostree-finalize-staged.service appears successful
I0417 18:21:30.720187    2551 daemon.go:1697] Current config: rendered-infra-5d1e3cebfddcee59ac7bde56152e0919
I0417 18:21:30.720192    2551 daemon.go:1698] Desired config: rendered-infra-a17773854499b91cb5cc86087b54cfe8
I0417 18:21:30.720202    2551 daemon.go:1706] state: Done

```

Some things to note
- If any of the changes result in a reboot action, all other policies will be ignored.
- There is no dedup of the final actions list. We don't expect users to do MC changes that causes multiple policies to be in effect for a single update very often, so this didn't seem like a pressing need. 
- There is a special exclusion for files defined in directory `/etc/containers/registries.d/` to account for some recent changes from https://github.com/openshift/machine-config-operator/pull/4160#discussion_r1548669673. In the future, the MCO plans to add directories and wildcard support. With that in place, this exception can be removed and made part of the standard policy.

djoshy commented:
/test unit
/test verify

djoshy commented:
/test e2e-gcp-op
/test e2e-gcp-op-techpreview

djoshy commented:
/retest-required

djoshy commented:
/retest-required

yuqi-zhang commented on manifests/machineconfigdaemon/clusterrole.yaml:
```diff
@@ -31,4 +34,4 @@ rules:
   - subjectaccessreviews
   verbs:
-  - create
+  - create
```
nit: I think we prefer to have a newline at the end of files

djoshy replied:
Will do!

yuqi-zhang commented on pkg/apihelpers/apihelpers.go:
```diff
@@ -8,8 +8,73 @@ import (
+				},
+			},
+			{
+				Path: "/var/lib/kubelet/config.json",
```
minor: given that these paths are referenced in a lot of places, maybe it's worth it to make them an operator-level constant.

Although maybe eventually we'd only want less references to these paths?

djoshy replied:
Will see if I can tighten these up!

yuqi-zhang commented on pkg/apihelpers/apihelpers.go:
```diff
@@ -8,8 +8,73 @@ import (
+	"k8s.io/klog/v2"
+)
+
+var (
```
I'm fine with it here but out of curiosity why did we decide to define this in apihelpers?

djoshy replied:
I sort of interpreted this as our internal MCO API specific helpers. I guess I could've put it in the regular helpers too, it just felt natural to put it there as NodeDisruption is an MCO only API object (:

yuqi-zhang commented on pkg/controller/common/helpers.go:
```diff
@@ -1008,6 +1010,48 @@ func CalculateConfigFileDiffs(oldIgnConfig, newIgnConfig *ign3types.Config) []st
+	for unit := range oldUnitSet {
+		_, ok := newUnitSet[unit]
+		if !ok {
+			// debug: remove
```
reminder to remove debug messages

djoshy replied:
Will do!

yuqi-zhang commented on pkg/daemon/constants/constants.go:
```diff
@@ -108,4 +108,7 @@ const (
 	CRIOServiceName = "crio"
+
+	// DaemonReloadServiceName is used to specify reloads and restarts of the systemd manager configuration
+	DaemonReloadServiceName = "daemon-reload"
```
nit: I guess this isn't technically a service. Maybe something like DaemonReloadCommand ?

yuqi-zhang commented on pkg/daemon/update.go:
```diff
@@ -83,6 +88,146 @@ func reloadService(name string) error {
+// If at any point an error occurs, we reboot the node so that node has correct configuration.
+func (dn *Daemon) performPostConfigChangeNodeDisruptionAction(postConfigChangeActions []opv1.NodeDisruptionPolicyStatusAction, configName string) error {
+
+	logSystem("Executing performPostConfigChangeNodeDisruptionAction(drain already complete/skipped) for config %s", configName)
```
`performPostConfigChangeNodeDisruptionAction` is certainly a mouthful :)

djoshy replied:
yeahhhh open to ideas lol

yuqi-zhang commented on pkg/daemon/update.go:
```diff
@@ -83,6 +88,146 @@ func reloadService(name string) error {
+	logSystem("Executing performPostConfigChangeNodeDisruptionAction(drain already complete/skipped) for config %s", configName)
+	for _, action := range postConfigChangeActions {
+		logSystem("Executing postconfig action: %v for config %s", action.Type, configName)
+		if action.Type == opv1.RebootStatusAction {
```
I wonder if we should make reboot/none a separate condition, i.e. if RebootStatusActions is in postConfigChangeActions we skip all other actions, and none can only exist if others do not?

Actually reading the rest of the PR, it looks like you have it set up such that if RebootStatusActions exists, is the list always just that one object so we are implicitly covered? (i.e. you can't have postConfigChangeActions contain both reboot and other actions like reload, and none shows up only if no other actions exist?)

djoshy replied:
Yeah the list of actions is built like this:
- if there is atleast one Reboot action, return a list of just a singular Reboot Action
- if there is a None action and other actions, strip out any None actions
- if there is only None actions in the list, return a list of just a singular None Action

Sorry if I made that a little convoluted, I tried to make the comments make sense!

yuqi-zhang commented on pkg/daemon/update.go:
```diff
@@ -474,6 +597,96 @@ func calculatePostConfigChangeActionFromFileDiffs(diffFileSet []string) (actions
+	for _, diffPath := range diffFileSet {
+		pathFound := false
+		for _, policyFile := range clusterPolicies.Files {
+			klog.Infof("comparing policy path %s to diff path %s", policyFile.Path, diffPath)
```
nit: we can probably merge some of the logging (although I guess since it's during an update, the logs should not happen that often). Up to you

djoshy replied:
Will see if I can tighten these up!

yuqi-zhang commented on pkg/daemon/update.go:
```diff
@@ -732,7 +979,48 @@ func (dn *Daemon) update(oldConfig, newConfig *mcfgv1.MachineConfig, skipCertifi
+	if fg != nil && fg.Enabled(v1.FeatureGateNodeDisruptionPolicy) {
+		mcop, err := dn.mcopClient.OperatorV1().MachineConfigurations().Get(context.TODO(), "cluster", metav1.GetOptions{})
+		if mcop == nil || apierrors.IsNotFound(err) {
+			return err
```
I'm wondering if in general we should have the backup (existing) path exist always just as a fallback in case the object doesn't exist for some reason.

For example, hypershift implementation has non-rebooting updates too but they probably won't have this object

Since this is tech preview and behind a FG I don't think we need to consider it now, but probably good to revisit for GA

djoshy replied:
Agree with making a card to fallback for GA. I'd guess a user would want to know how it failed and why? I guess we could log it somehow that we are in the fallback path, but maybe we can have this go up the object status when we get that working

djoshy replied:
Made a card for this: https://issues.redhat.com/browse/MCO-1146

yuqi-zhang commented on pkg/daemon/update.go:
```diff
@@ -732,7 +979,48 @@ func (dn *Daemon) update(oldConfig, newConfig *mcfgv1.MachineConfig, skipCertifi
+		if mcop == nil || apierrors.IsNotFound(err) {
+			return err
+		}
+		// TODO: wait for mcop.Status.NodeDisruptionPolicyStatus to populate, otherwise, error out?
```
I suppose its fine to have a check for whether clusterpolicies is empty, since it shouldn't if the MCO has populated it (maybe fallback?)

djoshy replied:
Yeah I wonder if there is a time window where clusterpolicies aren't ready yet, but the feature gate is on. We wouldn't want to immediately fall back in that case, so perhaps a timeout? I haven't seen this in my testing, it'd have to be a very small edge case

yuqi-zhang reviewed (commented):
Overall looks good! Some comments/questions inline but I think none of it needs to be blocking.

Thanks for the great work @djoshy

yuqi-zhang commented on pkg/daemon/update.go:
```diff
@@ -732,7 +979,48 @@ func (dn *Daemon) update(oldConfig, newConfig *mcfgv1.MachineConfig, skipCertifi
+	var actions []string
+	// If FeatureGateNodeDisruptionPolicy is set, calculate NodeDisruptionPolicy based actions for this MC diff
+	if fg != nil && fg.Enabled(v1.FeatureGateNodeDisruptionPolicy) {
+		mcop, err := dn.mcopClient.OperatorV1().MachineConfigurations().Get(context.TODO(), "cluster", metav1.GetOptions{})
```
I realized in https://github.com/openshift/machine-config-operator/pull/4281/  you added this as "MCOOperatorKnobsObjectName". We should eventually merge the two :)

djoshy replied:
ah good catch, I'll make sure to pick it up on next pass!

yuqi-zhang reviewed (approved):
Latest round of changes lgtm, going to wait for QE pre-merge testing as well before lgtm'ing

djoshy commented:
Rebased off master, and added a comment explaing why we are using ApplyConfiguration for the initial MachineConfiguration object

rioliu-rh commented:
/hold

rioliu-rh commented:
@djoshy found an issue when applying the featuregate [OCPBUGS-32511](https://issues.redhat.com/browse/OCPBUGS-32511)

djoshy commented:
Thanks for testing Sergio! I think this is a race condition during feature gate change, I've extended timeout and added a fallback to legacy path if the timeout is exceeded so we shouldn't degrade. Let me know if that works better

djoshy commented:
Made some additional fixes related to the transition to TechPreview featureset. In my tests, I noticed that operator would not get far enough to actually render the `NodeDisruptionPolicyStatus` as it was "stuck" in a mid feature gate state while the entire cluster was transitioning. Meanwhile the nodes would get assigned a new config to switch to; but with the 10 minute timeout I placed earlier, it takes them a long time to transition to the new config as it has to first hit the timeout and then fall back to execute the legacy update path. To clean up this UX, I did the following:

- moved up `syncMachineConfiguration` in the operator's sync functions so `NodeDisruptionPolicyStatus` is rendered earlier in the operator lifecycle
- dropped the timeout back to 2 minutes, after which the daemon will execute the old update path, not respecting node disruption policies. 

With the above fixes, the only variable to account for is the operator's lease acquire time, after which the status will be populated, with which the daemons can orchestrate updates. The two minute timeout is largely arbitrary. During the "feature gate" transition update, some nodes may have to fallback to the legacy path in case the operator is not fast enough to populate the `NodeDisruptionPolicyStatus`. Once the cluster has gone through the "feature gate" transition update, all future MC updates should respect the node disruption policy of the cluster.

Note: This is only applicable for cases where the feature gate was applied after the cluster was installed with a "Default" featureset. If the cluster featureset was TechPreview at install time, this issue would not happen, it is the transition that gunks up the process.

sinnykumari commented on pkg/daemon/update.go:
```diff
@@ -474,6 +598,96 @@ func calculatePostConfigChangeActionFromFileDiffs(diffFileSet []string) (actions
+		for _, policyFile := range clusterPolicies.Files {
+			klog.V(4).Infof("comparing policy path %s to diff path %s", policyFile.Path, diffPath)
+			if policyFile.Path == diffPath {
+				klog.Infof("NodeDisruptionPolicy found for diff file %s!", diffPath)
```
nit: Perhaps we should exclude `!` in the end since it can be confusing that it is part of filename.

djoshy replied:
ack, will fix alongside anything QE might find on next pass. Thanks!

sinnykumari commented on pkg/daemon/update.go:
```diff
@@ -83,6 +89,146 @@ func reloadService(name string) error {
+// If at any point an error occurs, we reboot the node so that node has correct configuration.
+func (dn *Daemon) performPostConfigChangeNodeDisruptionAction(postConfigChangeActions []opv1.NodeDisruptionPolicyStatusAction, configName string) error {
+
+	logSystem("Executing performPostConfigChangeNodeDisruptionAction(drain already complete/skipped) for config %s", configName)
```
Can we just say "Performing PostConfig Action".
We can skip `drain already complete/skipped` since this is already captured in https://github.com/openshift/machine-config-operator/blob/master/pkg/daemon/update.go#L793 if we drain.

sinnykumari reviewed (commented):
Few minor nits which are optional.
Other than that this looks great. Nice work David!
/lgtm

rioliu-rh commented:
pre-merge testing result:
verified following functionalities
- files actions
   - none
   - reboot
   - restart
   - reload
   - drain
   - daemon reload
   - multi-actions
- units actions
   - none
   - reboot
   - multi-actions
- sshkey actions
   - none
   - reboot
   - multi-actions

Issues: 
- [OCPBUGS-32511](https://issues.redhat.com/browse/OCPBUGS-32511) NodeDisruptionPolicyStatus was not ready context deadline exceeded
- [OCPBUGS-32739](https://issues.redhat.com/browse/OCPBUGS-32739) MachineConfigurations is only effective with name <cluster>
- [OCPBUGS-32783](https://issues.redhat.com/browse/OCPBUGS-32783) NodeDisruptionPolicy action reload cannot take effect

Test cases are linked to epic [MCO-507](https://issues.redhat.com/browse/MCO-507)
No merge blocker. issues can be fixed new PRs. 

/label qe-approved
/unhold

djoshy commented:
/override ci/prow/e2e-hypershift

Hypershift failures are unrelated and being fixed. Since the PR has passed it in the past, I'm overriding it.

sinnykumari commented:
Looks like https://github.com/openshift/hypershift/pull/3938 doesn't help much with HyperShift failure. Looking at recent failure on this PR doesn't look like MCO specific. Also, we had a successful run on this PR with current changes in https://prow.ci.openshift.org/view/gs/test-platform-results/pr-logs/pull/openshift_machine-config-operator/4267/pull-ci-openshift-machine-config-operator-master-e2e-hypershift/1782558388421398528 . This PR should be safe to go in. overriding e2e-hypershift test to  get this feature work merged.
/override ci/prow/e2e-hypershift
